### PR TITLE
feat: Added new mount property

### PR DESCRIPTION
**Pull request by atribecalledkwest**
This would replace the current automounting of bower_components folders. As it stands, NodeCG's current system isn't very good
- It isn't completely apparent to new users that they _should_ use bower if they are looking at an existing project, or that NodeCG does this automatic bower_components mounting in the first place.
- It pushes people towards one particular package manager
- Users who want to use other package managers have to work a bit harder than normal to get their directories mounted.

This would change that by allowing users to define a `mount` array in their package.json file that specifies an object containing both a `directory` and `endpoint` property.

A few things I want to add before this even is considered for merging:
- [x] Move error-checking code into bundle-parser
- [ ] More stringent error checking to inputs, as it stands I'm not sure how well this fares with odd inputs for relative files/mount endpoints, or if this would even need any of that.
- [ ] Tests for valid/invalid mountpoints.
- [x] Better error logging for when a mountpoint is skipped.

Other things I'd like inputs on:
- How should this be made as accessible/easily known to new users as possible? Make this option known in generator-nodecg? More documentation?

### CODE REVIEW

File: lib/components.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+const app = express();
+const path = require('path');
+const log = require('./logger')('nodecg/lib/components');
+const Bundles = require('./bundle-manager');
+
+log.trace('Adding Express routes for bundle components');
+
+const all = Bundles.all();
+
+for (let i = 0; i < all.length; i++) {
```
Comment (Lange) [inline lib/components.js]: I usually prefer to use `<array>.forEach` instead of `for` loops, where possible.

File: lib/components.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	const bundle = all[i];
+	if (bundle.mount) {
+		for (let j = 0; j < bundle.mount.length; j++) {
+			const mountdir = bundle.mount[j];
+			if (typeof mountdir.endpoint !== 'string' || typeof mountdir.directory !== 'string') {
```
Comment (Lange) [inline lib/components.js]: This validation (and any additional validation) should go somewhere in [`lib/bundle-parser`](https://github.com/nodecg/nodecg/tree/master/lib/bundle-parser). I'd probably make a new file for it called `mount.js` or `mounts.js`.

File: lib/components.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+const log = require('./logger')('nodecg/lib/components');
+const Bundles = require('./bundle-manager');
+
+log.trace('Adding Express routes for bundle components');
+
+const all = Bundles.all();
+
+for (let i = 0; i < all.length; i++) {
+	const bundle = all[i];
+	if (bundle.mount) {
```
Comment (Lange) [inline lib/components.js]: By inverting this condition to be `if (!bundle.mount || bundle.mount.length <= 0)` (or something), and then `continue;`ing (or `return;`ing if using `Array.forEach`), the code currently inside the `if` block can be moved left by one level of indentation, making it a bit nicer to read. I usually take this stylistic approach whenever I can. For example:

```js
foo.forEach(bar => {
    if (!bar.baz) {
        // Nothing for us to do then.
        return;
    }

    // Do something with baz.
    foo.bar.baz();
});
```

**Comment by Lange** [comment]
Just checking in -- did you want some help with the remaining 2 items on the task checklist in the first post?

**Comment by atribecalledkwest** [comment]
Yeah kinda, I could easily do any of the error checking stuff, but I am stumped when it comes to doing tests.

**Comment by Lange** [comment]
Sure, I can try to find some time to help write some scaffolding that would help you get started. Feel free to poke me on gitter if it's been a while, I'm juggling a lot right now so I might just forget.

**Comment by Lange** [comment]
Finally had some time to look at this. I've added tests and docs, and I think this is ready to merge.

@atribecalledkwest can you give everything a once-over and let me know if you're ready for me to merge this? It'll be included in the next release. Thanks, and I apologize for the really long delay in getting to this. Just been way too busy.

**Comment by atribecalledkwest** [comment]
It's fine, just gimme a bit, I'll give it a look over when I have a tiny bit more free time.

**Comment by atribecalledkwest** [comment]
AFAIK everything looks good, feel free to merge whenever.

**Pull request merged**